design/downstream-crl-design.md

```diff
@@ -0,0 +1,128 @@
+  routes:
+    - services:
+        - name: service
+          port: 80
+```
+
+## Detailed Design
+
+### CRL Secret
+
+The same approach shall be followed for configuring revocation lists as is used currently to store the CA certificates for client validation
+
+The CRL is stored in an opaque Kubernetes secret.
+The secret will be stored in the same namespace as the corresponding `HTTPProxy` object.
+The secret object shall contain entry named `crl.pem`.
+The constents shall be the CRL in PEM format.
+The file may contain "PEM bundle", that is, a list of CRLs concatenated in single file.
+
+Example:
+
+```bash
+kubectl create secret generic crl-for-client-validation --from-file=./crl.pem
+```
+
+### httpproxy.DownstreamValidation additions
+
+```go
+// DownstreamValidation defines how to verify the client certificate.
+type DownstreamValidation struct {
+  ...
+
+	// Name of a Kubernetes Opaque secret that contains a concatenated list of
+  // pem encoded crls.
+	// This will be used to verify that a client certificate has not been revoked
+	// +optional
+	// +kubebuilder:validation:MinLength=1
+	CACertificateRevocationList string `json:"caCrlSecret,omitempty"`
+}
+```
+
+### Envoy Configuration
+
+The new fields from `spec.virtualhost.tls.clientValidation` must be parsed and mapped to `auth.CommonTlsContext.ValidationContextType`
+
+- `spec.virtualhost.tls.clientValidation.caCrlSecret` -> `envoy_v3_tls.CommonTlsContext_ValidationContext.Crl`
+  - `internal.envoy.v3.auth.validationContext()` will be updated to accept a `crl` the same way it accepts a `ca`
+  - If empty it will not set the `crl` field, maintaining backward compatibility
+
+### Secret validation
+
+Currently basic validation is done on Opaque secrets with a `ca.crt` key, just to make sure that the length is non-zero.
+This validation would also be performed for the CRL.
+
+Validation is also performed to make sure that the CA Bundle has the correct PEM header and is of type `CERTIFICATE`.
+The same validation would be performed, but checking for type `CRL`
+
+## Alternatives Considered
+
+N/A
+
+## Security Considerations
+
+Secret must be in the same namespace as the HTTPProxy, similar requirements to the CA when using client cert authentication
+
+## Compatibility
+
+This change should be additive, so there should be no compatibility issues
+
+## Implementation
+
+E2E tests required:
+  * Happy path
+    * 1 CRL to one certificate
+    * 2+ CRL to 2+ certificates (same length)
+  * Errors
+    * Mismatch in CRL/Cert chain length
+
+## Open Issues
+
+- K8s secrets are limited in size, so there will be an inherent limitation to the CRL bundle in this design
+- How to handle an error where a user supplies a CRL for one certificate in a chain, but not all. In which case Envoy will fail to verify
+  - Short of decoding both the certificate list and CRL list, and making sure one exists for both, I don't have another answer. This seems out of scope for Contour here.
```

Comment:
Are you referring to the case which the parameter `only_verify_leaf_cert_crl` addresses (at some level) in the future?  Yes, I agree it is outside of scope for Contour for now - the CRL check is done at level where Contour is not actively involved.